cpp/src/dlpack/dlpack.cpp

```diff
@@ -0,0 +1,262 @@
+
+}  // namespace
+
+namespace detail {
+
+std::unique_ptr<experimental::table> from_dlpack(
+    DLManagedTensor const* managed_tensor,
+    rmm::mr::device_memory_resource* mr,
+    cudaStream_t stream)
+{
+  CUDF_EXPECTS(nullptr != managed_tensor, "managed_tensor is null");
+  auto const& tensor = managed_tensor->dl_tensor;
+
+  // We can copy from host or device pointers
+  CUDF_EXPECTS(kDLGPU == tensor.ctx.device_type ||
+               kDLCPU == tensor.ctx.device_type ||
+               kDLCPUPinned == tensor.ctx.device_type, 
+               "DLTensor must be GPU, CPU, or pinned type");
+
+  // Make sure the current device ID matches the Tensor's device ID
+  if (tensor.ctx.device_type != kDLCPU) {
+    int device_id = 0;
+    CUDA_TRY(cudaGetDevice(&device_id));
+    CUDF_EXPECTS(tensor.ctx.device_id == device_id, 
+                 "DLTensor device ID must be current device");
+  }
+
+  // Currently only 1D and 2D tensors are supported
+  CUDF_EXPECTS(tensor.ndim > 0 && tensor.ndim <= 2,
+               "DLTensor must be 1D or 2D");
+
+  CUDF_EXPECTS(tensor.shape[0] > 0, "DLTensor first dim empty");
+  CUDF_EXPECTS(tensor.shape[0] < std::numeric_limits<size_type>::max(),
+               "DLTensor first dim exceeds size supported by cudf");
+  if (tensor.ndim > 1) {
+    CUDF_EXPECTS(tensor.shape[1] > 0, "DLTensor second dim empty");
+    CUDF_EXPECTS(tensor.shape[1] < std::numeric_limits<size_type>::max(),
+               "DLTensor second dim exceeds size supported by cudf");
+  }
+
+  size_t const num_columns = (tensor.ndim == 2) ?
+    static_cast<size_t>(tensor.shape[1]) : 1;
+
+  // Validate and convert data type to cudf
+  data_type const dtype = DLDataType_to_data_type(tensor.dtype);
+
+  size_t const byte_width = size_of(dtype);
+  size_t const num_rows = static_cast<size_t>(tensor.shape[0]);
+  size_t const bytes = num_rows * byte_width;
+
+  // For 2D tensors, if the strides pointer is not null, then strides[1] is the
+  // number of elements (not bytes) between the start of each column
+  size_t const col_stride = (tensor.ndim == 2 && nullptr != tensor.strides) ?
+    byte_width * tensor.strides[1] : byte_width * num_rows;
+
+  auto tensor_data = reinterpret_cast<uintptr_t>(tensor.data) + tensor.byte_offset;
+
+  // Allocate columns and copy data from tensor
+  std::vector<std::unique_ptr<column>> columns(num_columns);
+  for (auto& col : columns) {
+    col = make_numeric_column(dtype, num_rows, UNALLOCATED, stream, mr);
+
+    CUDA_TRY(cudaMemcpyAsync(col->mutable_view().head<void>(),
+      reinterpret_cast<void*>(tensor_data), bytes, cudaMemcpyDefault, stream));
+
+    tensor_data += col_stride;
+  }
+
+  return std::make_unique<experimental::table>(std::move(columns));
+}
+
+DLManagedTensor* to_dlpack(table_view const& input,
+    rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+{
+  auto const num_rows = input.num_rows();
+  auto const num_cols = input.num_columns();
+  if (num_rows == 0) {
+    return nullptr;
```

Comment:
Is this well understood behavior?  Perhaps the documentation should reflect that if num_rows is 0 the return will be nullptr.

Reply:
Good point. I could alternatively return a tensor with the rows set to zero (or whatever a valid DLPack tensor is supposed to look like). I'll see which would be preferred

Reply:
Updated the docs to reflect this